src/snapred/backend/recipe/Recipe.py

```diff
@@ -0,0 +1,111 @@
+        self.mantidSnapper = utensils.mantidSnapper
+
+    ## Abstract methods which MUST be implemented
+
+    @abstractmethod
+    def chopIngredients(self, ingredients: Ingredients):
+        """
+        Chops off the needed elements of the ingredients.
+        """
+
+    @abstractmethod
+    def unbagGroceries(self, groceries: Dict[str, WorkspaceName]):
+        """
+        Unpacks the workspace data from the groceries.
+        """
+
+    @abstractmethod
+    def queueAlgos(self):
+        """
+        Queues up the procesing algorithms for the recipe.
+        Requires: unbagged groceries and chopped ingredients.
+        """
+
+    # methods which MAY be kept as is
+
+    def validateInputs(self, ingredients: Ingredients, groceries: Dict[str, WorkspaceName]):
+        """
+        Validate the input properties before chopping or unbagging
+        """
+        # cast the ingredients into the Ingredients type
+        try:
+            # to run the same as Ingredients.parse_obj(ingredients)
+            get_args(self.__orig_bases__[0])[0].parse_obj(ingredients)
+        except ValidationError as e:
+            raise e
+        # ensure all of the given workspaces exist
+        # NOTE may need to be tweaked to ignore output workspaces...
+        for ws in groceries.values():
+            if not self.mantidSnapper.mtd.doesExist(ws):
+                raise RuntimeError(f"The indicated workspace {ws} not found in Mantid ADS.")
+
+    def stirInputs(self):
+        """
+        Validation of chopped and unbagged inputs, to ensure consistency
+        Requires: unbagged groceries and chopped ingredients.
+        """
+        pass
+
+    def prep(self, ingredients: Ingredients, groceries: Dict[str, str]):
+        """
+        Convinience method to prepare the recipe for execution.
+        """
+        self.validateInputs(ingredients, groceries)
+        self.unbagGroceries(groceries)
+        self.chopIngredients(ingredients)
+        self.stirInputs()
+
+        self.queueAlgos()
+
+    def execute(self):
+        """
+        Final step in a recipe, executes the queued algorithms.
+        Requires: queued algorithms.
+        """
+        self.mantidSnapper.executeQueue()
+        return True
+
+    def cook(self, ingredients: Ingredients, groceries: Dict[str, str]) -> bool:
+        """
+        Main interface method for the recipe.
+        Given the ingredients and groceries, it prepares, executes and returns the final workspace.
+        """
+        self.prep(ingredients, groceries)
+        return self.execute()
+
+    def cater(self, shipment: List[Pallet]) -> bool:
+        """
+        A secondary interface method for the recipe.
+        It is a batched version of cook.
+        Given a shipment of ingredients and groceries, it prepares, executes and returns the final workspaces.
+        """
+        for ingredient, grocery in shipment:
+            self.prep(ingredient, grocery)
+        return self.execute()
```

Comment:
would this mean its up the the implementation of execute to accrew the batched results?

Reply:
If a recipe needs results, then it has to overwrite this.  See what I did in the `ApplyWhateverItWas` recipe

Reply:
It would be way better if it handled outputs automatically, I just wasn't sure how to do that